### PR TITLE
Generate EFI grub images for x86 platforms

### PR DESCRIPTION
**Pull request by sduponch**
- Add option in make menuconfig to build an EFI Grub image with legacy grub support.
- Update the x86 Makefile to generate an EFi grub image [ MBR | EFI | GrubModule | RootFs ]
- EFI image generate an UUID GPT signature to replace the too small legacy part signature.
- EFI image use the grub search files module to find the boot partition (as grub does not have a GPT uuid part search support).
- Fork grub build with efi platform support via a common.mk Makefile include to not fork the whole grub package and patches (host need both of legacy and efi modules to setup the bootloader on efi images).
- Add gptfdisk tools package to get sgdisk on host side, used for converting MBR to GPT tables.
- Fork gen_image_generic.sh in gen_image_efi.sh to add EFI related additional's partitions.
- Update platform.sh to support GPT UUID root signature in /proc/cmdline and use awk (that is already embeded in pivot root) to reorder bytes of the dumped partition signature).
- Create a kmod-efi-vars kernel module (modules.mk) to add EFI related Kernel config options that are required to boot with a console on HDMI (FB).
- Generate VDI and VMDK EFI images

Known limitations : 
- Sysupgrade -p is forced for EFI/GPT images.

```
GPT EFI image partition scheme
+--------------------------------------------------------------------------------------------------------------------------+
| MBR (raw 1MB) | Kernel (ext4 4MB) | EF00 (vfat 1MB) | EF02 (raw 1MB) | Rootfs (ext4/squash XX MB)  | GTP table (raw 16KB)|
+--------------------------------------------------------------------------------------------------------------------------+
```

Tested on : 
- Virtualbox legacy mode
- Virtualbox EFI mode
- Intel Nuc DE3815TYKE legacy mode
- Intel Nuc DE3815TYKE EFI mode
- Giada F-210 with an EFI only bios
- Intel Cherry trail platform with an EFI only bios

Not tested on x86 32bits platform

### CODE REVIEW

**Review by alive4ever — COMMENTED** [review]
tools/gptfdisk/Makefile: This should be on a separate pull request, since it introduces a new tool / base package.

File: tools/gptfdisk/Makefile [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#
+# Copyright (C) 2015 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=gptfdisk
```
Comment (alive4ever) [inline tools/gptfdisk/Makefile]: I mean this change and related changes, which introduces ``gdisk`` should be submitted as separate pull request.

File: target/linux/x86/image/Makefile [hunk 2]
```diff
@@ -10,6 +10,9 @@ include $(INCLUDE_DIR)/image.mk
 export PATH=$(TARGET_PATH):/sbin
 
 GRUB2_MODULES = biosdisk boot chain configfile ext2 linux ls part_msdos reboot serial vga
+GRUB2_MODULES_LEGACY = $(GRUB2_MODULES)
+GRUB2_MODULES_LEGACY += part_gpt part_msdos search fat exfat ext2
+GRUB2_MODULES_EFI = boot chain configfile ext2 linux ls part_msdos reboot serial part_gpt part_msdos search fat exfat ext2 efi_gop efi_uga gfxterm
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: There is no font module here. Does the grub framebuffer work, i.e. displays kernel text message?

File: target/linux/x86/image/Makefile [hunk 3]
```diff
@@ -217,14 +325,22 @@ define Image/Build
 	$(call Image/Build/$(1))
   ifneq ($(1),iso)
 	$(call Image/Build/grub2,$(1))
+	$(call Image/Build/efi,$(1))
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: This doesn't depend on ``ifeq ($(1),ext4)``. This means the UEFI images will be generated for both ext4 and squashfs image.

Does the generated squashfs image boot on EFI system?

I put the conditional ext4 so that the UEFI image is generated only for ext4 filesystem since I can't get the squashfs image to boot on EFI system.

File: package/boot/grub2/grub2-efi/Makefile [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+$(eval $(call HostBuild))
+$(eval $(call BuildPackage,grub2-efi))
+
+include $(INCLUDE_DIR)/kernel.mk
+
+define KernelPackage/efi-vars
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=EFI VARS modules
+  DEPENDS:=@EFI_IMAGES
+  KCONFIG:= \
```
Comment (alive4ever) [inline package/boot/grub2/grub2-efi/Makefile]: These kernel configurations should be included in the subtarget config (``target/linux/x86/64/config-default`` and ``target/linux/x86/generic/config-default``).

To modify the kernel configuration for subtarget, launch ``make CONFIG_TARGET=subtarget kernel_menuconfig``.

File: tools/gptfdisk/patches/000-gdisk-only.patch [hunk 5]
```diff
@@ -0,0 +1,13 @@
+diff --git a/Makefile b/Makefile
+index 8c8f9fd..8ea9718 100644
+--- a/Makefile
++++ b/Makefile
+@@ -11,7 +11,7 @@ MBR_LIB_OBJS=$(MBR_LIBS:=.o)
+ LIB_HEADERS=$(LIB_NAMES:=.h)
+ DEPEND= makedepend $(CXXFLAGS)
+ 
+-all:	cgdisk gdisk sgdisk fixparts
```
Comment (alive4ever) [inline tools/gptfdisk/patches/000-gdisk-only.patch]: This doesn't matter at all, since ``gdisk`` is host only package.

As I know, ``gdisk`` is intended for interactive use. For scripting and automation, ``sgdisk`` should be used.

File: config/Config-images.in [hunk 6]
```diff
@@ -224,6 +224,14 @@ menu "Target Images"
 		help
 		  If you don't know, 5 seconds is a reasonable default.
 
+	config EFI_IMAGES
+		bool "Build EFI GRUB images (Linux x86 or x86_64 host only)"
+		depends on TARGET_x86_64 || (TARGET_x86 && !TARGET_x86_rdc)
+		depends on TARGET_ROOTFS_EXT4FS || TARGET_ROOTFS_ISO || TARGET_ROOTFS_JFFS2 || TARGET_ROOTFS_SQUASHFS
+		select GRUB_IMAGES
+		select PACKAGE_grub2-efi
```
Comment (alive4ever) [inline config/Config-images.in]: No need to introduce a new package here.

File: package/boot/grub2/grub2-efi/Makefile [hunk 7]
```diff
@@ -0,0 +1,46 @@
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+
+PKG_NAME:=grub-efi
```
Comment (alive4ever) [inline package/boot/grub2/grub2-efi/Makefile]: No need for new package here.

**Review by alive4ever — CHANGES_REQUESTED** [review]
Here are my review, since the changes are rather too big for a single commit.

File: target/linux/x86/image/Makefile (hunk 2 above)
Comment (sduponch) [inline target/linux/x86/image/Makefile]: Good catch, the kernel messages are not displayed until the system load the framebuffer module.

File: target/linux/x86/image/Makefile (hunk 3 above)
Comment (sduponch) [inline target/linux/x86/image/Makefile]: Yes, the squashfs image boot correctly with EFI because the kernel partition is in ext4, only the rootfs partition is in squashfs (because squashfs is read only with an overlay in jffs2).

File: package/boot/grub2/grub2-efi/Makefile (hunk 4 above)
Comment (sduponch) [inline package/boot/grub2/grub2-efi/Makefile]: I think, the best place to do this is in target/linux/x86/modules.mk . We don't really need to modify the config-default to only support EFI and we don't want to append EFI stuff in a non EFI system.

File: package/boot/grub2/grub2-efi/Makefile (hunk 7 above)
Comment (sduponch) [inline package/boot/grub2/grub2-efi/Makefile]: A separate package could be avoided if build variant were supported for Host packages :\

So, i made a common.mk Makefile inspired from the toolchain/gcc to ease the maintainability of these grub packages. I will take a look on how you did that :)

File: config/Config-images.in (hunk 6 above)
Comment (sduponch) [inline config/Config-images.in]: Do we need an EFI platform grub package on EFI target (this package is not for the host side) ?

File: tools/gptfdisk/patches/000-gdisk-only.patch (hunk 5 above)
Comment (sduponch) [inline tools/gptfdisk/patches/000-gdisk-only.patch]: i did not success to use sgdisk :\

File: target/linux/x86/image/Makefile [hunk 8]
```diff
@@ -66,6 +70,92 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
...
+	mcopy -s -i "$(KDIR)/kernel.efi" $(KDIR)/grub2.efi/* ::/
+
+	# Append EFI boot image to MBR grub image
+	dd if="$(KDIR)/kernel.efi" of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 oflag=append conv=notrunc
+
+	# Append 33 block for GPT at end of the disk image
+	dd if="/dev/zero" of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=33 oflag=append conv=notrunc
+
+	# Set the new EFI partition (data are already present on image) and change PARTUUID of the boot partition
+	(echo n; echo 3; echo ; echo ; echo EF00; echo x; echo c; echo 2; echo $(EFI_SIGNATURE); echo w; echo Y) | gdisk "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: sgdisk equivalent

Creating partition 3 with type EF00 (EFI System Partition).
``sgdisk -n 3 -t 3:EF00 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

Changing UUID of partition 3 with predefined value.
``sgdisk -u 3:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

File: target/linux/x86/image/Makefile [hunk 9]
```diff
@@ -66,6 +70,92 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
...
+
+	# Append Grub EFI modules partition
+	dd if=/dev/zero of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=2048 oflag=append conv=notrunc
+	(echo x; echo e; echo m; echo n; echo 4; echo ; echo ; echo EF02; echo w; echo Y; echo Y) | gdisk "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: sgdisk equivalent

Moving second header (backup mbr) to the end of disk.
``sgdisk -e "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

Creating partition 4 as bios boot partition
``sgdisk -n 4 -t 4:EF02 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

File: tools/gptfdisk/patches/000-gdisk-only.patch (hunk 5 above)
Comment (alive4ever) [inline tools/gptfdisk/patches/000-gdisk-only.patch]: I added ``sgdisk`` equivalent for those actions.

**Review by alive4ever — COMMENTED** [review]
I'll run a build test to check whether this single commit works as expected.

File: target/linux/x86/image/Makefile [hunk 10]
```diff
@@ -66,6 +70,92 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
...
+		-d $(STAGING_DIR_HOST)/lib/grub/i386-pc \
+		-o $(KDIR)/grub2/core.img \
+		-O i386-pc \
+		-p '(hd0,gpt1)/boot/grub' \
+		-c $(KDIR)/root.grub/boot/grub/grub.cfg \
+		$(GRUB2_MODULES_LEGACY)
+	$(CP) $(STAGING_DIR_HOST)/lib/grub/i386-pc/*.img $(KDIR)/grub2/
+
+	# In order to append the EFI partitions we must pad the image so TARGET_IMAGES_PAD is forced
+	PADDING="y" SIGNATURE="$(SIGNATURE)" PATH="$(TARGET_PATH)" ./gen_image_generic.sh \
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: Instead of forcing ``y`` as value, I think it's better to make use of kconfig to select ``CONFIG_TARGET_IMAGES_PAD`` as ``EFI_IMAGES`` dependency.

File: config/Config-images.in [hunk 11]
```diff
@@ -224,6 +224,14 @@ menu "Target Images"
 		help
 		  If you don't know, 5 seconds is a reasonable default.
 
+	config EFI_IMAGES
+		bool "Build EFI GRUB images (Linux x86 or x86_64 host only)"
+		depends on TARGET_x86_64 || (TARGET_x86 && !TARGET_x86_rdc)
+		depends on TARGET_ROOTFS_EXT4FS || TARGET_ROOTFS_ISO || TARGET_ROOTFS_JFFS2 || TARGET_ROOTFS_SQUASHFS
+		select GRUB_IMAGES
```
Comment (alive4ever) [inline config/Config-images.in]: On ``target/linux/x86/image/Makefile``, there is a statement that ``TARGET_IMAGES_PAD`` is forced.

``TARGET_IMAGES_PAD`` can be selected here instead of forcing ``TARGET_IMAGES_PAD=y``

**Review by alive4ever — COMMENTED** [review]
Why not use kconfig to define ``TARGET_IMAGES_PAD`` instead of forcing ``TARGET_IMAGES_PAD=y``?

File: config/Config-images.in (hunk 11 above)
Comment (sduponch) [inline config/Config-images.in]: Selecting TARGET_IMAGES_PAD in CONFIG will affect size of the generated non the EFI images...

File: target/linux/x86/image/Makefile (hunk 10 above)
Comment (sduponch) [inline target/linux/x86/image/Makefile]: We need to force image PADDING here, because we append new partitions to the image if we set in CONFIG others images will be padded too.

File: target/linux/x86/image/Makefile (hunk 9 above)
Comment (sduponch) [inline target/linux/x86/image/Makefile]: Use of sgdisk add libpopt-dev as host dependency.
 sgdisk -n 3 -t 3:EF00 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
require use of the -g parameter to create GPT table

Comment (sduponch) [inline target/linux/x86/image/Makefile]: - sgdisk -u 3:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
+ sgdisk -u 2:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
GPT Signature must be on rootfs partition

File: target/linux/x86/image/Makefile (hunk 2 above)
Comment (sduponch) [inline target/linux/x86/image/Makefile]: I am not sure if grub font are really required, i did not see any difference ... ?

File: config/Config-images.in (hunk 11 above)
Comment (sduponch) [inline config/Config-images.in]: Thanks a lot for your review.

I have updated the Makefiles to use sgdisk and moved the kmod efi-vars from grub2-efi to modules.mk.

File: tools/gptfdisk/Makefile (hunk 1 above)
Comment (sduponch) [inline tools/gptfdisk/Makefile]: Submit a separate pull request to add a tools that is not used (yet) ? 
I can make a separate commit only for the tools, but i don't think the PR will be merged if the tool is not used...

File: config/Config-images.in (hunk 11 above)
Comment (sduponch) [inline config/Config-images.in]: PR splitted in two commit : 
- For sgdisk tools : https://github.com/openwrt/openwrt/pull/391/commits/3871e7c5c1e5e3beff8127e45d7d2f70dff174d7
- For EFI Grub : https://github.com/openwrt/openwrt/pull/391/commits/a9b15e43862459f45c75edf10aac5f82e150043a

File: target/linux/x86/image/Makefile [hunk 12]
```diff
@@ -66,6 +70,94 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
...
+		-O x86_64-efi \
+		-p /efi/boot \
+		-c $(KDIR)/grub2.efi/efi/boot/grub.cfg \
+		$(GRUB2_MODULES_EFI)
+
+	# Create the EFI vfat part image
+	rm $(KDIR)/kernel.efi || true
+	mkfs.fat -C $(KDIR)/kernel.efi -S 512 $(shell echo $$(( $(CONFIG_TARGET_KERNEL_PARTSIZE) * 1024 )))
+	mcopy -s -i "$(KDIR)/kernel.efi" $(KDIR)/grub2.efi/* ::/
+
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: How about moving this step to replace the first boot partition generation process, so that the first partition is EFI system partition (EF00), the second partition is bios boot (EF02), and the third partition is the root partition?.

This is important, because there will be a lot of space at the end of the disk after performing ``dd`` of the grub-efi-gpt image. With root partition placed on the last block, it is possible to expand the root partition to fill up entire free space without messing up another partition (EFI and bios boot).

File: target/linux/x86/image/Makefile [hunk 13]
```diff
@@ -66,6 +70,94 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
...
+	dd if="/dev/zero" of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=33 oflag=append conv=notrunc
+
+	# Set the new EFI partition (data are already present on image) and change PARTUUID of the boot partition
+	sgdisk -g -n 3 -t 3:EF00 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
+	sgdisk -u 2:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
+
+	# Append Grub EFI modules partition
+	dd if=/dev/zero of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=2048 oflag=append conv=notrunc
+	sgdisk -e "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
+	sgdisk -n 4 -t 4:EF02 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: The BIOS boot partition should be placed before the root partition, so it's possible to expand the root partition to fill entire free space.

```
+-----------------------------+----------------------------+----------------+
| EFI System Partition (EF00) | BIOS boot partition (EF02) | Root partition |
+-----------------------------+----------------------------+----------------+
```

**Review by alive4ever — COMMENTED** [review]
How about placing the root partition at the end so that it's possible to expand the root partition without messing with another partition afterwards?

File: target/linux/x86/image/Makefile (hunk 13 above)
Comment (sduponch) [inline target/linux/x86/image/Makefile]: Why not, we can try this :)

Comment (sduponch) [inline target/linux/x86/image/Makefile]: i have started to work here: 
- https://github.com/sduponch/openwrt/pull/2/files

For now, the image build and grub boot in legacy and efi mode but system fail to mount rootfs because of a bad geometry (block count exceeds size of device)

**Comment by sduponch** [comment]
> How about placing the root partition at the end so that it's possible to expand the root partition without messing with another partition afterwards?
Done :)

File: target/linux/x86/image/Makefile [hunk 14]
```diff
@@ -10,6 +10,9 @@ include $(INCLUDE_DIR)/image.mk
 export PATH=$(TARGET_PATH):/sbin
 
 GRUB2_MODULES = biosdisk boot chain configfile ext2 linux ls part_msdos reboot serial vga
+GRUB2_MODULES_LEGACY = $(GRUB2_MODULES)
+GRUB2_MODULES_LEGACY += part_gpt part_msdos search fat exfat ext2
```
Comment (alive4ever) [inline target/linux/x86/image/Makefile]: ``part_msdos`` and ``ext2`` mentioned twice here.

**Review by alive4ever — COMMENTED** [review]
Duplicate modules mentioned on ``GRUB2_MODULES_LEGACY`` variable

**Comment by sduponch** [comment]
Done. I have also changed "gen_image_efi.sh" to output MBR image and be able to setup the legacy grub before converting this partition to GPT (should improve hybrid compatibility)

**Comment by sduponch** [comment]
+---------------------+-----------------------+----------------------+--------------------- +-----------------------------+
| MBR (raw 1MB) | Kernel (ext4 1MB) | EF00 (vfat 1MB) | EF02 (raw 1MB) | Rootfs (ext/sq XX MB)  |
+---------------------+-----------------------+----------------------+----------------------+-----------------------------+

**Comment by alive4ever** [comment]
So fast!

That's all my review.

Let's hope that OpenWRT/LEDE developers got the guts to decide whether to commit the changes or discard these altogether, so that we can move on to another task.

File: package/boot/grub2/common.mk [hunk 15]
```diff
@@ -1,33 +1,24 @@
...
 PKG_VERSION:=2.02~beta2
 PKG_RELEASE:=1
 
-PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE:=grub-$(PKG_VERSION).tar.xz
 PKG_SOURCE_URL:=http://alpha.gnu.org/gnu/grub \
 	http://gnualpha.uib.no/grub/ \
 	http://mirrors.fe.up.pt/pub/gnu-alpha/grub/ \
 	http://www.nic.funet.fi/pub/gnu/alpha/gnu/grub/
 PKG_MD5SUM:=be62932eade308a364ea4bbc91295930
```
Comment (alive4ever) [inline package/boot/grub2/common.mk]: One more forgotten review if you don't mind.

I suggest replacing ``PKG_MD5SUM`` with ``PKG_HASH``.
``PKG_HASH`` is just SHA256 hash, which should be very easy to generate.

Comment (sduponch) [inline package/boot/grub2/common.mk]: Done

File: tools/gptfdisk/Makefile [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+PKG_VERSION:=1.0.1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=http://www.rodsbooks.com/gdisk/
+PKG_MD5SUM:=d7f3d306b083123bcc6f5941efade586
```
Comment (alive4ever) [inline tools/gptfdisk/Makefile]: This section should also be converted to ``PKG_HASH``

Comment (sduponch) [inline tools/gptfdisk/Makefile]: Done

**Comment by sduponch** [comment]
I have fixed the grub2-efi "--with-suffix" warning (the option is --program-suffix) and updated Makefile.
This allow to have no footprint on MBR image because we fully separate the two grub binaries.

**Comment by sduponch** [comment]
Backported to LEDE: https://github.com/lede-project/source/pull/1026

**Comment by sduponch** [comment]
Closing because outdated by the LEDE port
